### ISSUE TITLE
WOW!!!

### ISSUE THREAD

**Issue by thundergreen** [mariohock/Chatty#1]
Hello you are amazing! just a question.. HOW to use it? :) 

could you ake a small write-up?

Your Biggest fan!

**Comment by mariohock** [comment]
Hi, thank you. :-D

First of all you need AppDaemon (look at https://appdaemon.readthedocs.io/en/latest/ for an install guide). Note that version 4 is required (because of asyncio). Then copy chatty.py and chatty.yaml in the apps folder. (Also configure your XMPP accounts in chatty.yaml)

Once you have AppDaemon running, take a look in the logs to see whether Chatty is correctly loaded. It might crash because it cannot find the slixmpp library. Then you have to install it. In my case (on a hassbian installation), I entered the appdaemon venv and installed it with: "pip3 install slixmpp". If you use docker, the probably is some kind of startup/install script where you can call pip3!?

If Chatty is up and running your Home-Assistent XMPP account should be shown as online. Send a message to it and look at the logs if something was received. Then try sending a message. For an initial test, the easiest way should be to open chatty.py in a text editor and modify the function on_incoming_message. Just return a string (instead of None), then your string should be answered to every message you send to your Home-Assistent XMPP account. (If you're not a programmer, just skip to next step.)

If all this works, you can send messages from every AppDaemon app with: self.call_service("notify/jabber", message="your message"). Or from Home-Assistant by producing a "NOTIFY_JABBER" event (e.g., from developer tools) with 'message: "your message"' in the data field.

Hope this helps.

**Comment by thundergreen** [comment]
I try to figure out what system dev tool to install to buid slixmpp.. i am neat to the solution :) I'll report as soon as it runs :) 

Do you have some real life examples for me how i can use your app?

**Comment by mariohock** [comment]
p.s. I just added a "help" command (with dummy implementation). But now you can just send "help" and will get an answer from Chatty.

**Comment by thundergreen** [comment]
2020-01-30 20:57:52.594306 INFO Chatty: Incoming: 'Hell', from 'thorsten@emevth.no-ip.biz/Pix-Art Messenger.99zH

it seems online :) i've to update my xhatty-py then :P

**Comment by thundergreen** [comment]
@mariohock are u on discord? would like to discuss a bit with you :) 
ot just send me your JID :) or write me via xmpp :P my <JID is here 

I am nt a programmer but would like to extend this bot a bit :) i would just need a bit examples and help and i would provide ideas :) I already used hubot but it's crappy to setup. so i like your app and would like something to interact with my home assistant (house) :) do you think U wanna go this way with me?

**Comment by mariohock** [comment]
Great! As for real live examples: Look at the new commit with the "help" function. Just use it as a template to implement your own functions. I just implemented two commands "home" and "gone" that turns on/off my heating/radiators (but the implementation is quite specific to my home and only a few lines of code).

Actually, I never used discord before. It's a chat, right? Maybe I'll look into it..

**Comment by thundergreen** [comment]
just write me via xmpp :) It works well with event in homeassistant too I will watch your code to figure out how to make it.. then i will create a "real" help file with all commands

well.. u did not push you commit with home and gone command example :) 

another issue: I get messages by twice each time :(

**Comment by mariohock** [comment]
I just wrote you on XMPP. And, yes of course: Patches (merge requests) welcome.

Yeah, usually with AppDaemon, if you change the code of your app, the app it is reloaded. But here it appears that the new version is started but the old one not stopped. I have to look into it if this is a bug in AppDaemon 4 or if my app doesn't get killed correctly. Restarting AppDaemon helps, for now.

**Comment by thundergreen** [comment]
I use for example node-red to send messages to homeassistant :) I created then two sensors and like this i can send messages on my dashboard.. even pics

see:


![image](https://user-images.githubusercontent.com/13320319/73486851-315c8300-43a6-11ea-9e01-296637becc18.png)

**Comment by mariohock** [comment]
Great, I would appreciate a real help file (especially from an outside perspective, and hass.io). Yes, the plan was to make my commands quite specific to my home, so not make it public.

**Comment by thundergreen** [comment]
if possible just add some examples and people can adapt / learn

**Comment by mariohock** [comment]
I pushed the "home" and "gone" thing as a branch, just as an example to see some working code. (Actually the commands are "heim" and "weg".)

**Comment by thundergreen** [comment]
cool! thanks a lot!

**Comment by mariohock** [comment]
The issue where Chatty doesn't die correctly is now fixed.

**Comment by thundergreen** [comment]
btw.. this is my appdaemon config to install and compile slixmpp:

```
{
  "system_packages": [
    "libjpeg",
    "tiff",
    "gcc",
    "python3-dev",
    "musl-dev"
  ],
  "python_packages": [
    "PyMySQL",
    "hass-apps",
    "adutils~=0.4.9",
    "slixmpp"
  ],
  "init_commands": [],
  "disable_auto_token": true,
  "log_level": "info"
}
```

**Comment by mariohock** [comment]
I added some error handling for the XMPP connection, but it's not really tested, yet. Also I extended the code examples, including the reading of states. 

This makes the code a little more complex, however, to understand how it works, one can still look at the initial commit and afterwards at the latest version.

Also the callbacks are now async, this means you have to put "async" in front of the "def" of your callbacks. Therefore, you now can call "await" within the callbacks.

**Comment by meisterlampe** [comment]
Cool. I just wanted to create an issue because the bot is not responding anymore after some hours (after the night in the morning), but now I'll try the exended version first. :)

**Comment by meisterlampe** [comment]
Chatty send a message this night, that it has recovered and indeed it is still responding.
So error handling for the XMPP connection works well. * thumbsup *

**Comment by mariohock** [comment]
Great, thanks for the conformation.

But why do you loose connection every night? You can, of course, disable this message if you don't want to be woken up every night (just comment out the respective line of code).

**Comment by thundergreen** [comment]
Would someone mind how I can use chatty as a home Assistant. Service to send me messages an pics ? Or appdaemon... My idea is having pics sent or even templates with entity states .. a reason life example would be highly appreciated as I think hat official. Component is dead

**Comment by mariohock** [comment]
I think the official component is able to send pictures (over XEP-0363: HTTP File Upload). Maybe one can use this code as a starting point to implement it in Chatty (since both use Slixmpp). But, of course, mind the copyright. Unfortunately, I won't have time for this myself, anytime soon. But patches for Chatty are very welcome.

Converting Chatty to a Home-Assistant service should be possible, as well. But, how knows, maybe there would be the same race conditions..?

**Comment by thundergreen** [comment]
yeah.. that would be awesome to send pics.. but how would an automation look like with your app? 

service: fire_event or something like this?

**Comment by thundergreen** [comment]
```
event_test:
  alias: event_test
  sequence:
  - event: NOTIFY_JABBER
    event_data:
      message: This is a message
```

This works :)
